File: src/main/java/net/axay/blueutils/geometry/Circle.java
package net.axay.blueutils.geometry;

import java.util.ArrayList;
import java.util.List;

public class Circle {

    public static List<Location2D> getCircle(int radius) {
        return getCircle(0, 0, radius);
    }

    @SuppressWarnings("DuplicatedCode")
    public static List<Location2D> getCircle(int x_loc, int y_loc, int radius) {

        List<Location2D> locationList = new ArrayList<>();

        while (radius >= 0) {

            int d = -radius;
            int x = radius;
            int y = 0;

            while (y <= x) {

                locationList.add(new Location2D(x_loc + x, y_loc + y));
                locationList.add(new Location2D(x_loc + x, y_loc - y));
                locationList.add(new Location2D(x_loc - x, y_loc + y));
                locationList.add(new Location2D(x_loc - x, y_loc - y));
                locationList.add(new Location2D(x_loc + y, y_loc + x));
                locationList.add(new Location2D(x_loc + y, y_loc - x));
                locationList.add(new Location2D(x_loc - y, y_loc + x));
                locationList.add(new Location2D(x_loc - y, y_loc - x));

                d = d + 2*y + 1;
                y = y + 1;

                if (d > 0) {
                    d = d - 2*x + 2;
                    x = x - 1;
                }

            }

            radius--;

        }

        return locationList;

    }

}

File: src/main/java/net/axay/blueutils/geometry/Circle.kt
package net.axay.blueutils.geometry

data class Circle(val radius: Int, val center: Location2D) {

    val fillLocations: Set<Location2D> by lazy {

        val locationList: MutableSet<Location2D> = HashSet()

        val xLoc: Int = center.x
        val yLoc: Int = center.y

        var currentRadius: Int = radius
        while (currentRadius >= 0) {

            var d: Int = -currentRadius
            var x: Int = currentRadius
            var y = 0

            while (y <= x) {

                locationList.addCircleLoc(x, y)
                locationList.addCircleLoc(x, -y)
                locationList.addCircleLoc(-x, y)
                locationList.addCircleLoc(-x, -y)
                locationList.addCircleLoc(y, x)
                locationList.addCircleLoc(y, -x)
                locationList.addCircleLoc(-y, x)
                locationList.addCircleLoc(-y, -x)

                d += 2 * y + 1
                y++

                if (d > 0) {
                    d += -2 * 3 + 2
                    x--
                }

            }

            currentRadius--

        }

        return@lazy locationList

    }

    private fun MutableSet<Location2D>.addCircleLoc(first: Int, second: Int) {
        this += Location2D(center.x + first, center.y + second)
    }

}